src/System.Numerics.Vectors/src/System/Numerics/Matrix3x2.cs

```diff
@@ -119,6 +119,22 @@ public static Matrix3x2 CreateTranslation(Vector2 position)
             return result;
         }
 
+        /// <summary>
+        /// Creates a translation matrix from the given vector.
+        /// </summary>
+        /// <param name="position">The translation position.</param>
+        /// <param name="result">A translation matrix.</param>
+        public static void CreateTranslation(in Vector2 position, out Matrix3x2 result)
+        {
+            result.M11 = 1.0f;
+            result.M12 = 0.0f;
+            result.M21 = 0.0f;
+            result.M22 = 1.0f;
+
+            result.M31 = position.X;
+            result.M32 = position.Y;
```

Comment:
Rather than duplicating all of these implementations, would it make sense to implement the existing overload in terms of the new one? e.g.
```C#
public static Matrix3x2 CreateTranslation(Vector2 position)
{
    Matrix3x2 result;
    CreateTranslation(in position, out result);
    return result;
}
```

Reply:
This will force architectures using `System V ABI` to have worse performance.

Reply:
> This will force architectures using System V ABI to have worse performance.

To what extent?  If significant, then ok, but there is a _ton_ of duplicated code being added here.

Reply:
If the Jit used `__vectorcall` then wouldn't need any `in` params for the Vector types as they are all register sized?

Reply:
@stephentoub, I believe it would be measurable. It should be whatever the latency/overhead of copying from register to memory would be and back would be.

Reply:
@benaadams, it depends on how many args you pass. IIRC, `__vectorcall` gives 6 SIMD registers and supports HVA type values (so it can pass `Matrix4x4` all in register, for example).

Reply:
> I believe it would be measurable

Can we measure then?

Reply:
I've seen various poor CQ issues in the JIT around handling the SysV conventions for structs; things may be better for Vectors/HVAs but I' be kind of surprised if this is the case. In particular the jit seems to spill promoted structs to memory before calls (then reloading into registers) and vice versa on method entry.

So, please look carefully at the actual code the jit generates in any sort of A/B experiment. The results may say more about the abilities or limitations of the jit than any inherent "best" way of passing args.

Reply:
If there are limitations or poor CQ with regard to Vectors/HVAs, then I would prefer to see those get fixed over adding new APIs which only patch the issue.

That being said, I will be sure to also look at the codegen when comparing.

Reply:
I've been able to run some benchmarks today.  I'm using the [suite that @mellinoe put together](https://gist.github.com/mellinoe/5af3ae7db013a6ab740fc8b80910b1b8) in the issue #157.

## Machine 1: OS=ubuntu 16.04 Processor=Intel Xeon CPU E5-1620 0 3.60GHz, ProcessorCount=8 .NET Core 2.1.0-preview1-25919-02 (Framework 4.6.25917.03), 64bit RyuJIT

### Baseline (none of my changes)

|                  Method |       Mean |     Error |    StdDev |
|------------------------ |-----------:|----------:|----------:|
|        Matrix4x4ByValue | 271.720 ns | 0.2741 ns | 0.2430 ns |
|        QuaterionByValue | 206.123 ns | 0.0605 ns | 0.0566 ns |
|          Vector3ByValue |  11.839 ns | 0.0379 ns | 0.0354 ns |
| Vector3SimpleAddByValue |   2.118 ns | 0.0017 ns | 0.0016 ns |

### Current changes (duplicated code)

|                  Method |        Mean |     Error |    StdDev |
|------------------------ |------------:|----------:|----------:|
|        Matrix4x4ByValue | 275.4029 ns | 0.5192 ns | 0.4335 ns |
|          Matrix4x4ByRef | 201.6788 ns | 0.9473 ns | 0.7910 ns |
|        QuaterionByValue | 202.4441 ns | 0.1351 ns | 0.1128 ns |
|          QuaterionByRef | 282.7828 ns | 0.1976 ns | 0.1848 ns |
|          Vector3ByValue |  12.3704 ns | 0.0042 ns | 0.0035 ns |
|            Vector3ByRef |   0.7947 ns | 0.0009 ns | 0.0007 ns |
| Vector3SimpleAddByValue |   1.8711 ns | 0.0018 ns | 0.0016 ns |
|   Vector3SimpleAddByRef |   0.8011 ns | 0.0211 ns | 0.0187 ns |

### Doing some refactoring suggested here
See [this commit](https://github.com/eerhardt/corefx/commit/9b9a89c2031399e2d18c4b3ea900f3d6343cff47) for the code changes

|                  Method |        Mean |     Error |    StdDev |
|------------------------ |------------:|----------:|----------:|
|        Matrix4x4ByValue | 250.2053 ns | 0.2409 ns | 0.2135 ns |
|          Matrix4x4ByRef | 235.3991 ns | 0.1826 ns | 0.1619 ns |
|        QuaterionByValue | 338.3308 ns | 0.3273 ns | 0.2733 ns |
|          QuaterionByRef | 283.7789 ns | 0.4899 ns | 0.4343 ns |
|          Vector3ByValue |  21.1915 ns | 0.0120 ns | 0.0106 ns |
|            Vector3ByRef |   0.7941 ns | 0.0013 ns | 0.0012 ns |
| Vector3SimpleAddByValue |   1.8746 ns | 0.0022 ns | 0.0019 ns |
|   Vector3SimpleAddByRef |   0.8000 ns | 0.0135 ns | 0.0120 ns |


## Machine 2: OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.19) Processor=, ProcessorCount=8 .NET Core 2.1.0-preview1-25907-02 (Framework 4.6.25901.06), 64bit RyuJIT

I also ran the same tests on my Windows machine:

### Baseline (none of my changes)

|                  Method |       Mean |     Error |    StdDev |
|------------------------ |-----------:|----------:|----------:|
|        Matrix4x4ByValue | 205.941 ns | 0.8044 ns | 0.7524 ns |
|        QuaterionByValue | 157.711 ns | 0.5429 ns | 0.5078 ns |
|          Vector3ByValue |   9.050 ns | 0.0350 ns | 0.0328 ns |
| Vector3SimpleAddByValue |   1.391 ns | 0.0098 ns | 0.0087 ns |

### Current changes (duplicated code)

|                  Method |        Mean |     Error |    StdDev |
|------------------------ |------------:|----------:|----------:|
|        Matrix4x4ByValue | 212.2870 ns | 1.2065 ns | 1.1285 ns |
|          Matrix4x4ByRef | 159.5631 ns | 0.5604 ns | 0.5242 ns |
|        QuaterionByValue | 158.7156 ns | 0.7299 ns | 0.6095 ns |
|          QuaterionByRef | 147.2884 ns | 0.6121 ns | 0.5111 ns |
|          Vector3ByValue |   9.0204 ns | 0.0310 ns | 0.0290 ns |
|            Vector3ByRef |   9.8258 ns | 0.0334 ns | 0.0296 ns |
| Vector3SimpleAddByValue |   1.3929 ns | 0.0118 ns | 0.0110 ns |
|   Vector3SimpleAddByRef |   0.4443 ns | 0.0075 ns | 0.0066 ns |

## Doing some refactoring suggested here

|                  Method |        Mean |     Error |    StdDev |
|------------------------ |------------:|----------:|----------:|
|        Matrix4x4ByValue | 181.4270 ns | 0.5213 ns | 0.4876 ns |
|          Matrix4x4ByRef | 172.6201 ns | 0.6639 ns | 0.5885 ns |
|        QuaterionByValue | 181.0366 ns | 0.7396 ns | 0.6918 ns |
|          QuaterionByRef | 148.7591 ns | 0.6140 ns | 0.5744 ns |
|          Vector3ByValue |  19.4549 ns | 0.0934 ns | 0.0828 ns |
|            Vector3ByRef |   9.2186 ns | 0.0154 ns | 0.0129 ns |
| Vector3SimpleAddByValue |   1.3787 ns | 0.0106 ns | 0.0099 ns |
|   Vector3SimpleAddByRef |   0.4396 ns | 0.0110 ns | 0.0097 ns |


----

On both Windows and Ubuntu, it appears that the `Matrix4x4ByValue` test gets better with the refactoring, but both `QuaterionByValue` and `Vector3ByValue` get worse.

My thoughts are that we shouldn't do the refactoring at this time, as it adds risk to this change


----

Another observation is that @tannergooding's initial primary concern holds for the `QuaterionByValue` vs. `QuaterionByRef` test on Ubuntu.

|                  Method |        Mean |     Error |    StdDev |
|------------------------ |------------:|----------:|----------:|
|        QuaterionByValue | 202.4441 ns | 0.1351 ns | 0.1128 ns |
|          QuaterionByRef | 282.7828 ns | 0.1976 ns | 0.1848 ns |

In this scenario, on my Ubuntu machine, using the new `ref` overloads appears to lead to poorer performance.  But on my Windows machine, this isn't true and it wasn't true for @mellinoe's [reports](https://github.com/dotnet/corefx/issues/157#issuecomment-232233031) either.

Reply:
> But on my Windows machine, this isn't true and it wasn't true for @mellinoe's reports either.

On Windows, only the default x64 calling convention is used and there isn't support for `__vectorcall` today. This means that (on Windows) all of these are passed as normal structs (requiring shadow copying to the stack) and will likely see a perf increase (due to no longer requiring shadow-copying of the values).

On Unix based systems, we use the System V ABI, which allows these values to be passed in register, without shadow copying to the stack. However, as @AndyAyersMS pointed out, the code gen for `System V` ABI isn't the best and we are still shadow copying the values in some scenarios (which appears to be why you are seeing improvements).

I'm still working on collecting actual codegen data here.